File: packages/react-native-amazon-connect-webrtc/__tests__/BackendClient.test.ts
import { assertSecureBaseUrl, BackendClient } from '../src/BackendClient';
import { AuthError, BackendError, InvalidRequestError, RateLimitedError } from '../src/errors';
import type { CallRequest } from '../src/types';

const REQUEST: CallRequest = {
  callType: 'audio',
  context: { issueType: 'billing', tier: 'gold' },
  device: {
    platform: 'iOS',
    osVersion: '17.5',
    appVersion: '1.0.0',
    deviceModel: 'iPhone15,2',
    locale: 'en-GB',
    networkType: 'wifi',
  },
};

const SESSION = {
  contactId: 'contact-1',
  participantId: 'participant-1',
  participantToken: 'ptoken-1',
  callType: 'audio',
  meeting: {
    meetingId: 'meeting-1',
    mediaRegion: 'eu-west-2',
    mediaPlacement: { audioHostUrl: 'wss://a', signalingUrl: 'wss://s' },
  },
  attendee: { attendeeId: 'attendee-1', joinToken: 'jtoken-1' },
};

function jsonResponse(status: number, body: unknown): Response {
  return new Response(JSON.stringify(body), {
    status,
    headers: { 'Content-Type': 'application/json' },
  });
}

function client(fetchFn: typeof fetch, token = 'jwt-123') {
  return new BackendClient(
    { baseUrl: 'https://api.example.com/v1', tokenProvider: () => token },
    { fetchFn, sleep: async () => undefined, random: () => 0.5 },
  );
}

describe('assertSecureBaseUrl', () => {
  it('accepts https and localhost http, rejects remote http', () => {
    expect(() => assertSecureBaseUrl('https://api.example.com/v1')).not.toThrow();
    expect(() => assertSecureBaseUrl('http://localhost:3000')).not.toThrow();
    expect(() => assertSecureBaseUrl('http://10.0.2.2:3000')).not.toThrow();
    expect(() => assertSecureBaseUrl('http://api.example.com/v1')).toThrow(BackendError);
    expect(() => assertSecureBaseUrl('not a url')).toThrow(BackendError);
  });
});

describe('startCall', () => {
  it('POSTs /calls with bearer token + idempotency key and returns the session', async () => {
    const fetchFn = jest.fn(async (url: RequestInfo | URL, init?: RequestInit) => {
      expect(String(url)).toBe('https://api.example.com/v1/calls');
      const headers = init?.headers as Record<string, string>;
      expect(headers.Authorization).toBe('Bearer jwt-123');
      expect(headers['Idempotency-Key']).toBe('key-1');
      expect(headers['X-Correlation-Id']).toBe('corr-1');
      expect(JSON.parse(String(init?.body))).toEqual(REQUEST);
      return jsonResponse(201, SESSION);
    }) as unknown as typeof fetch;

    const session = await client(fetchFn).startCall(REQUEST, 'key-1', 'corr-1');
    expect(session.contactId).toBe('contact-1');
    expect(session.attendee.joinToken).toBe('jtoken-1');
  });

  it('omits the Authorization header when the token is empty', async () => {
    const fetchFn = jest.fn(async (_url: RequestInfo | URL, init?: RequestInit) => {
      const headers = init?.headers as Record<string, string>;
      expect('Authorization' in headers).toBe(false);
      return jsonResponse(201, SESSION);
    }) as unknown as typeof fetch;

    await client(fetchFn, '').startCall(REQUEST, 'key-1');
  });

  it('retries 429 and 5xx reusing the SAME idempotency key, then succeeds', async () => {
    const keys: string[] = [];
    let call = 0;
    const fetchFn = jest.fn(async (_url: RequestInfo | URL, init?: RequestInit) => {
      keys.push((init?.headers as Record<string, string>)['Idempotency-Key']!);
      call += 1;
      if (call === 1) return jsonResponse(429, { error: { code: 'RATE_LIMITED', message: 'slow down' } });
      if (call === 2) return jsonResponse(502, { error: { code: 'UPSTREAM_ERROR', message: 'connect down' } });
      return jsonResponse(201, SESSION);
    }) as unknown as typeof fetch;

    await client(fetchFn).startCall(REQUEST, 'key-stable');
    expect(keys).toEqual(['key-stable', 'key-stable', 'key-stable']);
  });

  it('gives up after maxAttempts with the last error', async () => {
    const fetchFn = jest.fn(async () =>
      jsonResponse(429, { error: { code: 'RATE_LIMITED', message: 'still throttled' } }),
    ) as unknown as typeof fetch;

    await expect(client(fetchFn).startCall(REQUEST, 'k')).rejects.toBeInstanceOf(RateLimitedError);
    expect(fetchFn).toHaveBeenCalledTimes(3);
  });

  it('does NOT retry 4xx and maps the error envelope', async () => {
    const fetchFn = jest.fn(async () =>
      jsonResponse(400, { error: { code: 'INVALID_CALL_TYPE', message: 'bad callType' } }),
    ) as unknown as typeof fetch;

    const promise = client(fetchFn).startCall(REQUEST, 'k');
    await expect(promise).rejects.toBeInstanceOf(InvalidRequestError);
    await expect(promise).rejects.toMatchObject({ code: 'INVALID_CALL_TYPE' });
    expect(fetchFn).toHaveBeenCalledTimes(1);
  });

  it('maps 401/403 to AuthError', async () => {
    const fetchFn = jest.fn(async () =>
      jsonResponse(403, { error: { code: 'FORBIDDEN', message: 'no token' } }),
    ) as unknown as typeof fetch;

    await expect(client(fetchFn).startCall(REQUEST, 'k')).rejects.toBeInstanceOf(AuthError);
  });

  it('retries network errors and surfaces BackendError when they persist', async () => {
    const fetchFn = jest.fn(async () => {
      throw new Error('ECONNRESET');
    }) as unknown as typeof fetch;

    await expect(client(fetchFn).startCall(REQUEST, 'k')).rejects.toBeInstanceOf(BackendError);
    expect(fetchFn).toHaveBeenCalledTimes(3);
  });
});

describe('participant connection + DTMF', () => {
  it('exchanges the participantToken for a connectionToken', async () => {
    const fetchFn = jest.fn(async (url: RequestInfo | URL, init?: RequestInit) => {
      expect(String(url)).toBe('https://api.example.com/v1/calls/connections');
      expect(JSON.parse(String(init?.body))).toEqual({ participantToken: 'ptoken-1' });
      return jsonResponse(201, { connectionToken: 'ctoken-1', expiry: 'later' });
    }) as unknown as typeof fetch;

    await expect(client(fetchFn).createParticipantConnection('ptoken-1')).resolves.toBe('ctoken-1');
  });

  it('sends DTMF digits and validates them client-side', async () => {
    const fetchFn = jest.fn(async (url: RequestInfo | URL, init?: RequestInit) => {
      expect(String(url)).toBe('https://api.example.com/v1/calls/dtmf');
      expect(JSON.parse(String(init?.body))).toEqual({ connectionToken: 'ctoken-1', digits: '1*#,' });
      return jsonResponse(200, { sent: true });
    }) as unknown as typeof fetch;

    const c = client(fetchFn);
    await c.sendDtmf('ctoken-1', '1*#,');
    await expect(c.sendDtmf('ctoken-1', 'abc')).rejects.toBeInstanceOf(InvalidRequestError);
    await expect(c.sendDtmf('ctoken-1', '')).rejects.toBeInstanceOf(InvalidRequestError);
    expect(fetchFn).toHaveBeenCalledTimes(1); // invalid digits never reach the network
  });
});

describe('endCall', () => {
  it('DELETEs /calls/{contactId} with the id URL-encoded', async () => {
    const fetchFn = jest.fn(async (url: RequestInfo | URL, init?: RequestInit) => {
      expect(String(url)).toBe('https://api.example.com/v1/calls/contact%2F1');
      expect(init?.method).toBe('DELETE');
      return new Response(null, { status: 204 });
    }) as unknown as typeof fetch;

    await client(fetchFn).endCall('contact/1');
  });
});

File: packages/flutter_amazon_connect_webrtc/test/backend_client_test.dart
import 'dart:convert';

import 'package:flutter_amazon_connect_webrtc/flutter_amazon_connect_webrtc.dart';
import 'package:flutter_test/flutter_test.dart';
import 'package:http/http.dart' as http;
import 'package:http/testing.dart';

CallRequest _req() => const CallRequest(
      callType: CallType.audio,
      device: DeviceInfo(platform: 'iOS', appVersion: '4.2.0'),
      context: {'issueType': 'billing'},
    );

const _okBody = {
  'contactId': 'c-1',
  'participantId': 'p-1',
  'participantToken': 'pt-1',
  'callType': 'audio',
  'meeting': {
    'meetingId': 'm-1',
    'mediaPlacement': {'audioHostUrl': 'https://a', 'signalingUrl': 'wss://s'},
  },
  'attendee': {'attendeeId': 'a-1', 'joinToken': 'jt-1'},
};

BackendClient _client(MockClient mock) => BackendClient(
      baseUrl: Uri.parse('https://api.test/v1'),
      tokenProvider: () async => 'jwt-token',
      httpClient: mock,
      sleep: (_) async {}, // no real delay in tests
    );

void main() {
  test('FR-F7: startCall posts to /v1/calls with bearer token and returns a CallSession', () async {
    late http.Request captured;
    final mock = MockClient((req) async {
      captured = req;
      return http.Response(jsonEncode(_okBody), 201);
    });

    final session = await _client(mock).startCall(_req(), idempotencyKey: 'k-1');

    expect(captured.method, 'POST');
    expect(captured.url.toString(), 'https://api.test/v1/calls');
    expect(captured.headers['Authorization'], 'Bearer jwt-token');
    expect(captured.headers['Idempotency-Key'], 'k-1');
    final sent = jsonDecode(captured.body) as Map<String, dynamic>;
    expect(sent['callType'], 'audio');
    expect(session.contactId, 'c-1');
    expect(session.attendee.joinToken, 'jt-1');
  });

  test('US-5: the SAME idempotency key is reused across retries (429 then 201)', () async {
    final keys = <String?>[];
    var calls = 0;
    final mock = MockClient((req) async {
      calls++;
      keys.add(req.headers['Idempotency-Key']);
      if (calls == 1) {
        return http.Response(jsonEncode({'error': {'code': 'RATE_LIMITED', 'message': 'slow'}}), 429);
      }
      return http.Response(jsonEncode(_okBody), 201);
    });

    final session = await _client(mock).startCall(_req(), idempotencyKey: 'k-42');

    expect(calls, 2);
    expect(keys, ['k-42', 'k-42']); // reused, not regenerated
    expect(session.contactId, 'c-1');
  });

  test('US-4: a 429 that never clears becomes a RateLimitedException', () async {
    final mock = MockClient((req) async =>
        http.Response(jsonEncode({'error': {'code': 'RATE_LIMITED', 'message': 'slow'}}), 429),);
    await expectLater(
      _client(mock).startCall(_req(), idempotencyKey: 'k'),
      throwsA(isA<RateLimitedException>()),
    );
  });

  test('US-3: a 401 becomes an AuthException and is not retried', () async {
    var calls = 0;
    final mock = MockClient((req) async {
      calls++;
      return http.Response(jsonEncode({'error': {'code': 'UNAUTHORIZED', 'message': 'no'}}), 401);
    });
    await expectLater(
      _client(mock).startCall(_req(), idempotencyKey: 'k'),
      throwsA(isA<AuthException>()),
    );
    expect(calls, 1); // auth failure is terminal, no retry
  });

  test('a 400 becomes an InvalidRequestException (no retry)', () async {
    final mock = MockClient((req) async =>
        http.Response(jsonEncode({'error': {'code': 'INVALID_CALL_TYPE', 'message': 'bad'}}), 400),);
    await expectLater(
      _client(mock).startCall(_req(), idempotencyKey: 'k'),
      throwsA(isA<InvalidRequestException>()),
    );
  });

  test('endCall issues DELETE /v1/calls/{contactId} and accepts 204', () async {
    late http.Request captured;
    final mock = MockClient((req) async {
      captured = req;
      return http.Response('', 204);
    });
    await _client(mock).endCall('contact-9');
    expect(captured.method, 'DELETE');
    expect(captured.url.toString(), 'https://api.test/v1/calls/contact-9');
    expect(captured.headers['Authorization'], 'Bearer jwt-token');
  });

  test('createParticipantConnection posts the participantToken and returns the connectionToken',
      () async {
    late http.Request captured;
    final mock = MockClient((req) async {
      captured = req;
      return http.Response(jsonEncode({'connectionToken': 'ct-1', 'expiry': 'x'}), 201);
    });
    final token = await _client(mock).createParticipantConnection('pt-1');
    expect(captured.url.toString(), 'https://api.test/v1/calls/connections');
    expect(jsonDecode(captured.body), {'participantToken': 'pt-1'});
    expect(token, 'ct-1');
  });

  test('sendDtmf posts connectionToken + digits to /calls/dtmf', () async {
    late http.Request captured;
    final mock = MockClient((req) async {
      captured = req;
      return http.Response(jsonEncode({'sent': true}), 200);
    });
    await _client(mock).sendDtmf(connectionToken: 'ct-1', digits: '1*#');
    expect(captured.url.toString(), 'https://api.test/v1/calls/dtmf');
    expect(jsonDecode(captured.body), {'connectionToken': 'ct-1', 'digits': '1*#'});
  });

  test('sendDtmf surfaces backend errors as typed exceptions', () async {
    final mock = MockClient((req) async =>
        http.Response(jsonEncode({'error': {'code': 'UPSTREAM_ERROR', 'message': 'x'}}), 502),);
    await expectLater(
      _client(mock).sendDtmf(connectionToken: 'ct', digits: '1'),
      throwsA(isA<BackendException>()),
    );
  });

  test('registerDevice posts the push token to /v1/devices', () async {
    late http.Request captured;
    final mock = MockClient((req) async {
      captured = req;
      return http.Response(jsonEncode({'customerId': 'cust-1', 'platform': 'iOS'}), 200);
    });
    await _client(mock).registerDevice(
      customerId: 'cust-1',
      platform: 'iOS',
      pushToken: 'voip-token',
    );
    expect(captured.method, 'POST');
    expect(captured.url.toString(), 'https://api.test/v1/devices');
    expect(jsonDecode(captured.body), {
      'customerId': 'cust-1',
      'platform': 'iOS',
      'pushToken': 'voip-token',
    });
    expect(captured.headers['Authorization'], 'Bearer jwt-token');
  });

  test('answerOutboundCall exchanges the callId for a CallSession', () async {
    late http.Request captured;
    final mock = MockClient((req) async {
      captured = req;
      return http.Response(jsonEncode(_okBody), 200);
    });
    final session = await _client(mock).answerOutboundCall('call-1');
    expect(captured.method, 'POST');
    expect(captured.url.toString(), 'https://api.test/v1/calls/outbound/call-1/answer');
    expect(session.contactId, 'c-1');
    expect(session.attendee.joinToken, 'jt-1');
  });

  test('answerOutboundCall surfaces 410 (no longer ringing) as InvalidRequestException', () async {
    final mock = MockClient((req) async => http.Response(
        jsonEncode({'error': {'code': 'CALL_NO_LONGER_RINGING', 'message': 'gone'}}), 410,),);
    await expectLater(
      _client(mock).answerOutboundCall('call-1'),
      throwsA(
        isA<InvalidRequestException>().having((e) => e.code, 'code', 'CALL_NO_LONGER_RINGING'),
      ),
    );
  });

  test('declineOutboundCall posts to .../decline and accepts 204', () async {
    late http.Request captured;
    final mock = MockClient((req) async {
      captured = req;
      return http.Response('', 204);
    });
    await _client(mock).declineOutboundCall('call-2');
    expect(captured.method, 'POST');
    expect(captured.url.toString(), 'https://api.test/v1/calls/outbound/call-2/decline');
  });
}
